### PR TITLE
Item: Sonic

### PR DESCRIPTION
**Pull request by genious97**

### CODE REVIEW

File: game/scripts/vscripts/items/sonic.lua [hunk 1]
```diff
@@ -0,0 +1,64 @@
+item_sonic = class(ItemBaseClass)
+
+LinkLuaModifier("modifier_item_sonic", "items/sonic.lua", LUA_MODIFIER_MOTION_NONE)
+LinkLuaModifier("modifier_sonic_fly", "modifiers/modifier_sonic_fly.lua", LUA_MODIFIER_MOTION_NONE)
+
+function item_sonic:GetIntrinsicModifierName()
+  return "modifier_sonic_fly"
```
Comment (chrisinajar) [inline game/scripts/vscripts/items/sonic.lua]: Intrinsic modifier is the one that's always on. You want "modifier_item_sonic" here.

File: game/scripts/vscripts/items/sonic.lua [hunk 2]
```diff
@@ -0,0 +1,64 @@
+item_sonic = class(ItemBaseClass)
```
Comment (chrisinajar) [inline game/scripts/vscripts/items/sonic.lua]: Sonic is a transformation item so you should use `TransformationBaseClass`

File: game/scripts/vscripts/items/sonic.lua [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+LinkLuaModifier("modifier_item_sonic", "items/sonic.lua", LUA_MODIFIER_MOTION_NONE)
+LinkLuaModifier("modifier_sonic_fly", "modifiers/modifier_sonic_fly.lua", LUA_MODIFIER_MOTION_NONE)
+
+function item_sonic:GetIntrinsicModifierName()
+  return "modifier_sonic_fly"
+end
+
+function item_sonic:OnToggle()
```
Comment (chrisinajar) [inline game/scripts/vscripts/items/sonic.lua]: Instead of making this a toggle item, just use the existing transformation item mechanics.

You can remove this whole function and replace it with
```lua
function item_sonic:GetTransformationModifierName()
  return "modifier_sonic_fly"
end
```

File: game/scripts/vscripts/items/sonic.lua [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+  if state == true then
+    caster.AddNewModifier(caster, self, "modifier_sonic_fly", {})
+  else
+    caster.RemoveModifierByName("modifier_sonic_fly")
+  end
+  ability:StartCooldown(ability:GetCooldown(ability:GetLevel()))
+end
+
+function item_sonic:OnOwnerDied()
```
Comment (chrisinajar) [inline game/scripts/vscripts/items/sonic.lua]: Modifiers are automatically removed when you die unless the modifiers says that it should stay active or is an intrinsic modifier. You probably needed this while testing because you accidentally set the fly to be intrinsic.

File: game/scripts/vscripts/modifiers/modifier_sonic_fly.lua [hunk 5]
```diff
@@ -0,0 +1,24 @@
+modifier_sonic_fly = class(ModifierBaseClass)
+
+function modifier_sonic_fly:DeclareFunctions()
+  return {
+      MODIFIER_PROPERTY_MOVESPEED_BONUS_CONSTANT,
+      MODIFIER_PROPERTY_BONUS_VISION_PERCENTAGE,
+      MODIFIER_PROPERTY_STATUS_RESISTANCE
```
Comment (chrisinajar) [inline game/scripts/vscripts/modifiers/modifier_sonic_fly.lua]: Have you tested if `MODIFIER_PROPERTY_STATUS_RESISTANCE` works? I hear rumors....

File: game/scripts/npc/items/custom/item_sonic.txt [hunk 6]
```diff
@@ -0,0 +1,92 @@
+"DOTAItems"
+{
+  //=================================================================================================================
+  // Recipe: Sonic
+  //=================================================================================================================
+  "item_recipe_sonic"
+  {
+    "ID"                  "8473"
```
Comment (chrisinajar) [inline game/scripts/npc/items/custom/item_sonic.txt]: You'll need to update these ID's since lots of new stuff made it in, such as electrician.

File: game/scripts/npc/items/custom/item_sonic.txt [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  //=================================================================================================================
+  // Sonic
+  //=================================================================================================================
+  "item_sonic"
+  {
+    "ID"                  "8474"
+    "BaseClass"           "item_lua"
+    "AbilityBehavior"			"DOTA_ABILITY_BEHAVIOR_IMMEDIATE | DOTA_ABILITY_BEHAVIOR_NO_TARGET | DOTA_ABILITY_BEHAVIOR_TOGGLE"
```
Comment (chrisinajar) [inline game/scripts/npc/items/custom/item_sonic.txt]: Make the type `DOTA_ABILITY_BEHAVIOR_NO_TARGET | DOTA_ABILITY_BEHAVIOR_IGNORE_CHANNEL`

File: game/scripts/vscripts/modifiers/modifier_sonic_fly.lua (hunk 5 above)
Comment (genious97) [reply game/scripts/vscripts/modifiers/modifier_sonic_fly.lua]: Upon further testing and research I am unable to find a way to change the status resistance

File: game/scripts/vscripts/items/sonic.lua (hunk 4 above)
Comment (Trildar) [reply game/scripts/vscripts/items/sonic.lua]: Intrinsic modifiers are still removed on death if they're not set to not RemoveOnDeath, but they get reapplied on respawn. Consequently, pretty much all intrinsic modifiers ought to be unpurgable and RemoveOnDeath false. Though the latter is typically only important in somewhat corner cases.

**Comment by warpdragon** [comment]
a random sword i made sonic

**Comment by genious97** [comment]
I believe I have fixed what you have requested.

**Comment by l34Um1** [comment]
When I try to test this I get a ton of LUA errors and then my game crashes.

File: game/scripts/vscripts/items/sonic.lua [hunk 8]
```diff
@@ -0,0 +1,48 @@
+item_sonic = class(TransformationBaseClass)
+
+LinkLuaModifier("modifier_item_sonic", "items/sonic.lua", LUA_MODIFIER_MOTION_NONE)
+LinkLuaModifier("modifier_sonic_fly", "modifiers/modifier_sonic_fly.lua", LUA_MODIFIER_MOTION_NONE)
+
+function item_sonic:GetIntrinsicModifierName()
+  return "modifier_item_sonic"
+end
+
+function item_sonic:GetTransformtionModifierName()
```
Comment (Trildar) [inline game/scripts/vscripts/items/sonic.lua]: Typo, "Transformtion".

File: game/scripts/npc/items/custom/item_sonic.txt [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    "ItemBaseLevel"       "1"
+    "UpgradesItems"       "item_sonic_2;item_sonic_3"
+
+    "ItemCost"            "6615"
+    "ItemShopTags"        "sonic"
+    "ItemQuality"         ""
+    "ItemAliases"         "sonic"
+    "ItemInitialCharges"  "-1"
+    "ItemDisplayCharges"  "0"
+    "ItemRequiresCharges" "0"
```
Comment (Trildar) [inline game/scripts/npc/items/custom/item_sonic.txt]: If the item doesn't use charges, then there's no need to even have these keys.

File: game/scripts/npc/items/custom/item_sonic.txt [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  //=================================================================================================================
+  "item_sonic"
+  {
+    "ID"                  "8502"
+    "BaseClass"           "item_lua"
+    "AbilityBehavior"			"DOTA_ABILITY_BEHAVIOR_NO_TARGET | DOTA_ABILITY_BEHAVIOR_IGNORE_CHANNEL"
+    "AbilityTextureName"  "custom/sonic"
+    "ScriptFile"          "items/sonic.lua"
+
+    "AbilityCooldown"     "8.0"
```
Comment (Trildar) [inline game/scripts/npc/items/custom/item_sonic.txt]: Needs a shared cooldown key to make all the levels share a cooldown, though I guess it doesn't matter much in this case. I don't remember what the key is called exactly, or where it's usually put. Refer to other item KVs.

File: game/scripts/vscripts/items/sonic.lua [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+modifier_item_sonic = class(ModifierBaseClass)
+
+function modifier_item_sonic:IsHidden()
+  return true
+end
+
+function modifier_item_sonic:IsPurgable()
+  return false
+end
```
Comment (Trildar) [inline game/scripts/vscripts/items/sonic.lua]: Should also have `RemoveOnDeath` function returning false since it's an intrinsic modifier.

File: game/scripts/vscripts/modifiers/modifier_sonic_fly.lua [hunk 12]
```diff
@@ -0,0 +1,24 @@
+modifier_sonic_fly = class(ModifierBaseClass)
+
+function modifier_sonic_fly:DeclareFunctions()
+  return {
+      MODIFIER_PROPERTY_MOVESPEED_BONUS_CONSTANT,
+      MODIFIER_PROPERTY_BONUS_VISION_PERCENTAGE,
+      --MODIFIER_PROPERTY_STATUS_RESISTANCE
+  }
```
Comment (Trildar) [inline game/scripts/vscripts/modifiers/modifier_sonic_fly.lua]: I think you should leave the status resistance code in. Though I'm unsure if it will support negative values even when Valve does fix it.

File: game/scripts/npc/npc_items_custom.txt [hunk 13]
```diff
@@ -65,6 +65,7 @@
 #base "items/custom/item_shroud_2.txt"
 #base "items/custom/item_shroud_3.txt"
 #base "items/custom/item_shroud_4.txt"
+#base "items/custom/item_sonic.txt"
```
Comment (Trildar) [inline game/scripts/npc/npc_items_custom.txt]: You should create the KV for the higher levels too.

**Comment by l34Um1** [comment]
Okay so the game doesn't crash anymore but I still get a ton of LUA errors and the item does nothing.

**Comment by chrisinajar** [comment]
This item almost certainly doesn't work and can't work until Valve fixes the API bugs they introduced (read: never). We should redesign this item to not interact with status resistance or simply close the PR and forget the idea until something changes.

**Comment by genious97** [comment]
So, would you like me to close this pr?

**Comment by l34Um1** [comment]
I'm okay with changing it to a simple damage amp like old MoM instead.

**Comment by genious97** [comment]
Okay :smile:

**Comment by carlosrpg** [comment]
Any update on this?

**Comment by genious97** [comment]
Sorry I haven't had time to work on this with school and exams. If anyone wants this that's fine.

**Comment by chrisinajar** [comment]
We've since fixed status resistance so we can do this the "right way" now.

**Comment by GymLeaderGiovanni** [comment]
Will it support negative values though?

**Comment by chrisinajar** [comment]
if not, it's already custom so should be easy to get working... some effects might act odd, perhaps, we'll see...

**Comment by chrisinajar** [comment]
It'd be great if someone could finish this up, it's super close to working :+1:
